# Wikipedia Notable Life Expectancies

# [Notebook 2 of 4: Data Cleaning](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_data_clean_thanak_2022_06_13.ipynb)

## Context

The


## Objective

The

### Data Dictionary

- Feature: Description

## Importing Necessary Libraries

In [1]:
# To structure code automatically
%load_ext nb_black

# To import/export sqlite databases
import sqlite3 as sql

# To help with reading and manipulating data
import pandas as pd
import numpy as np
import re

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)
# To define the maximum number of rows to be displayed in a dataframe
pd.set_option("display.max_rows", 200)

# To supress warnings
# import warnings

# warnings.filterwarnings("ignore")

# To set some visualization attributes
pd.set_option("max_colwidth", 150)

<IPython.core.display.Javascript object>

## Data Overview

### Reading, Sampling, and Checking Data Shape

In [2]:
# Reading the wp_life_expect_raw_complete dataset
conn = sql.connect("wp_life_expect_raw_complete.db")
data = pd.read_sql("SELECT * FROM wp_life_expect_raw_complete", conn)

# Making a working copy
df = data.copy()

# Checking the shape
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns.")

# Checking first 2 rows of the data
df.head(2)

There are 133900 rows and 6 columns.


month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21  
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12

<IPython.core.display.Javascript object>

In [3]:
# Checking last 2 rows of the data
df.tail(2)

month_year day           name  \
133898  June 2022   9  Oleg Moliboga   
133899  June 2022   9       Zou Jing   

                                                                         info  \
133898    , 69, Russian volleyball player, Olympic champion (1980) and coach.   
133899  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  
133898        https://en.wikipedia.org/wiki/Oleg_Moliboga              2  
133899  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3

<IPython.core.display.Javascript object>

In [4]:
# Checking a sample of the data
df.sample(5)

month_year day                   name  \
78460  September 2015  20       Jagmohan Dalmiya   
99112  September 2018  11  Richard Newbold Adams   
77638       July 2015  31       W. Eugene Wilson   
1977    November 1994   8        Marianne Straub   
96782        May 2018   8    Frøystein Wedervang   

                                                                                                                                                        info  \
78460  , 75, Indian cricket official, President of International Cricket Council (1997–2000) and Board of Control for Cricket in India (2001–2004), cardi...   
99112                                                                                                                         , 94, American anthropologist.   
77638                                                                                                                             , 86, American politician.   
1977                                                                                                                         , 85, British textile designer.   
96782                                                                                                                            , 100, Norwegian economist.   

                                                         link num_references  
78460          https://en.wikipedia.org/wiki/Jagmohan_Dalmiya             34  
99112     https://en.wikipedia.org/wiki/Richard_Newbold_Adams              2  
77638          https://en.wikipedia.org/wiki/W._Eugene_Wilson              3  
1977            https://en.wikipedia.org/wiki/Marianne_Straub             14  
96782  https://en.wikipedia.org/wiki/Fr%C3%B8ystein_Wedervang              1

<IPython.core.display.Javascript object>

#### Observations:
- There are 133,900 rows and 6 columns.
- `month_year` contains the month and year of death, while `day` contains the day of the month of death.
- `name` is the notable person's name.  It is a nominal feature that will not be used for analysis, but will be maintained for any referencing needs.
- `info` contains multiple items including the notable person's "age, country of citizenship at birth, subsequent country of citizenship (if applicable), reason for notability, (and) cause of death (if known)."
- `link` is the url to the notable person's individual Wikipedia page.  If such a page does not exist, there is either a non-working link (https://en.wikipedia.orgNone), or the link is to a page with a message that the page does not exist for that individual.  `link` is a unique identifier for all entries, except the 6 with the non-working link, which do have unique `name` values from each other.
- `num_references` contains the number of references on the notable person's individual Wikipedia page.  This feature serves as a proxy measure of notability.
- Prior to EDA, our task will be to extract the individual elements that are comined in `month_year` and `info` columns.

### Checking Data Types, Duplicates, and Null Values

In [5]:
# Checking data types and null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133894 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

In [6]:
# Checking duplicate rows
df.duplicated().sum()

0

<IPython.core.display.Javascript object>

In [7]:
# Check percentage of null values by column
df.isnull().sum() / df.count() * 100

month_year        0.000000
day               0.000000
name              0.004481
info              0.000000
link              0.000000
num_references    0.000000
dtype: float64

<IPython.core.display.Javascript object>

In [8]:
# Checking number of missing values per row (not necessary here, but done to keep process standard)
df.isnull().sum(axis=1).value_counts()

0    133894
1         6
dtype: int64

<IPython.core.display.Javascript object>

#### Observations:
- Our dataset was saved to and read from the database without any hiccups.
- As expected, we have 6 entries that are missing `name`, but we will find it in their `info` values.
- All columns are currently of object type.  We will need to appropriately typecast them after separating the information in `month_year` and `info`.

## Data Cleaning

### Addressing Missing `name` Values

In [9]:
# Checking rows with missing name values
missing_name = df[df["name"].isna()]
missing_name

month_year day  name  \
18937      August 2001  11  None   
24985     January 2004  22  None   
27458       March 2005   1  None   
34077        July 2007  11  None   
64769   September 2013  29  None   
106613  September 2019  12  None   

                                                                                      info  \
18937   Kevin Kowalcyk, 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                Vincent Palmer, 37, British criminal.   
27458                                             Barry Stigler, 57, American voice actor.   
34077                                          Nana Gualdi, 75, German singer and actress.   
64769                                 Scott Workman, 47, American stuntman (, , ), cancer.   
106613                                                Thami Shobede, 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- These rows vary from the main set as there is a substring containing the person's name at the start of the `info` string.
- As there are so few rows missing `name`, let us address this issue first.

In [10]:
# For loop to copy name value from info value and remove name from info value
treat_rows = missing_name.index
for i in treat_rows:
    info = df.loc[i, "info"]
    info_lst = info.split(sep=",", maxsplit=1)

    name = info_lst[0].strip()
    df.loc[i, "name"] = name
    df.loc[i, "info"] = re.sub(name, "", info).strip()

# Re-check rows
df.loc[treat_rows, :]

month_year day            name  \
18937      August 2001  11  Kevin Kowalcyk   
24985     January 2004  22  Vincent Palmer   
27458       March 2005   1   Barry Stigler   
34077        July 2007  11     Nana Gualdi   
64769   September 2013  29   Scott Workman   
106613  September 2019  12   Thami Shobede   

                                                                        info  \
18937   , 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                , 37, British criminal.   
27458                                            , 57, American voice actor.   
34077                                       , 75, German singer and actress.   
64769                                , 47, American stuntman (, , ), cancer.   
106613                                               , 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- Missing `name` values have been addressed and those names have been removed from `info` values.

In [11]:
# Re-check info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133900 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

#### Observations:
- We have no remaining missing values.
- Let us treat the `month_year` column next.

### Separating `month` and `year`

In [12]:
# Separating month and year into 2 columns and typecasting year as integer
df.loc[:, "year"] = df["month_year"].apply(lambda x: x.split(sep=" ")[1].strip())
df["year"] = df["year"].apply(lambda x: int(x))

df.loc[:, "month"] = df["month_year"].apply(lambda x: x.split(sep=" ")[0])
df.head(2)

month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month  
0  1994  January  
1  1994  January

<IPython.core.display.Javascript object>

In [13]:
# Dropping month_year column
df.drop("month_year", axis=1, inplace=True)
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month  
0  1994  January  
1  1994  January

<IPython.core.display.Javascript object>

### Treating `info`

#### Checking a Sample

In [14]:
# Checking a sample of info
df.sample(5)

day             name  \
88355  27     Sin Kek Tong   
33970  28  Catherine Troeh   
37900   6       Karl Kuehl   
32971  23     John Ritchie   
89689   6    John Bennison   

                                                                                         info  \
88355                  , 72, Singaporean politician, founder of the Singapore People's Party.   
33970                                    , 96, American native people activist and historian.   
37900  , 70, American baseball scout, coach and manager (Montreal Expos), pulmonary fibrosis.   
32971                 , 65, British footballer for Stoke City F.C., club's record goalscorer.   
89689                                              , 92, Australian businessman (Wesfarmers).   

                                                                     link  \
88355                          https://en.wikipedia.org/wiki/Sin_Kek_Tong   
33970                       https://en.wikipedia.org/wiki/Catherine_Troeh   
37900                            https://en.wikipedia.org/wiki/Karl_Kuehl   
32971  https://en.wikipedia.org/wiki/John_Ritchie_(footballer,_born_1941)   
89689                         https://en.wikipedia.org/wiki/John_Bennison   

      num_references  year     month  
88355              9  2017  February  
33970              7  2007      June  
37900              3  2008    August  
32971              9  2007  February  
89689              7  2017       May

<IPython.core.display.Javascript object>

#### Observations:
- First, let us check for any rows that are missing digits, and therefore the age target, within `info` and remove them.
- Also, it would be helpful to remove information contained within parentheses, as we will not be using it.

#### Checking and Dropping Rows Lacking Digits (and therefore Age Data) within `info`

In [15]:
# For loop to extract index of rows without digits in info value
remove_lst = []
for index in df.index:
    pattern = r"\d"
    if re.search(pattern, df.loc[index, "info"]) is None:
        remove_lst.append(index)
print(len(remove_lst), "rows")
df.loc[remove_lst, :].sample(5)

925 rows


day                          name  \
70403   17             Fariha al-Berkawi   
18252   13             Kinfe Gebremedhin   
54491   10                George Beattie   
108472  14                Gita Siddharth   
19620   11  Dasht-e Qaleh Taliban ambush   

                                                            info  \
70403                                 , Libyan politician, shot.   
18252   , Ethiopian Chief of Security and Immigration, murdered.   
54491                                     , Scottish footballer.   
108472                                    , Indian actress (, ).   
19620                               Journalists killed in the \n   

                                                                                                link  \
70403                                                https://en.wikipedia.org/wiki/Fariha_al-Berkawi   
18252                                                https://en.wikipedia.org/wiki/Kinfe_Gebremedhin   
54491                                      https://en.wikipedia.org/wiki/George_Beattie_(footballer)   
108472                                                  https://en.wikipedia.org/wiki/Gita_Siddharth   
19620   https://en.wikipedia.org/wiki/Timeline_of_the_War_in_Afghanistan_(2001%E2%80%93present)#2001   

       num_references  year     month  
70403               6  2014      July  
18252               5  2001       May  
54491               4  2012     March  
108472              6  2019  December  
19620              86  2001  November

<IPython.core.display.Javascript object>

In [16]:
# Dropping rows missing age data, resetting index, and checking new shape of df
df.drop(remove_lst, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132975, 7)

<IPython.core.display.Javascript object>

#### Observations:
- 925 rows were removed as they lacked the target data for `age`.

#### Removing Information within Parentheses in `info`

In [17]:
# Regular expression for parenthesis and its contents
pattern = r"\(.*\)"

# Subbing empty string for parentheses and stripping white space
df.loc[:, "info"] = df["info"].apply(lambda x: re.sub(pattern, "", x).strip())
df.loc[[21850, 125430], :]

day                 name  \
21850   11  Sir Michael Clapham   
125430  23       Jean-Luc Nancy   

                                                                                                      info  \
21850   , 90, British industrialist, president of the Confederation of British Industry from 1972 to 1974.   
125430                                                                           , 81, French philosopher.   

                                                                 link  \
21850   https://en.wikipedia.org/wiki/Michael_Clapham_(industrialist)   
125430                   https://en.wikipedia.org/wiki/Jean-Luc_Nancy   

       num_references  year     month  
21850              14  2002  November  
125430              8  2021    August

<IPython.core.display.Javascript object>

#### Observation:
- Parentheses and information within has been removed from `info`.
- Next, we will follow the Wikipedia-defined fields to divide the `info` values.

#### Splitting `info` on Commas into Separate Columns

In [18]:
# For loop to split info on commas and separate into respective new columns and removing leading/trailing white space and periods
for i, item in enumerate(df["info"]):
    info_lst = item.split(",")

    for j in range(len(info_lst)):
        df.loc[i, f"info_{j}"] = info_lst[j].strip(" .")

# Checking the first 2 rows
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_0 info_1           info_2                        info_3  \
0  1994  January            86   British dancer  ballet designer and director   
1  1994  January            68  Irish economist                        writer   

         info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  
0           NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
1  and academic    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

In [19]:
# Checking the last 2 rows
df.tail(2)

day           name  \
132973   9  Oleg Moliboga   
132974   9       Zou Jing   

                                                                         info  \
132973          , 69, Russian volleyball player, Olympic champion  and coach.   
132974  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132973        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132974  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_0 info_1                     info_2  \
132973  2022  June            69  Russian volleyball player   
132974  2022  June            86           Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132973                   Olympic champion  and coach    NaN    NaN    NaN   
132974  member of the Chinese Academy of Engineering    NaN    NaN    NaN   

       info_7 info_8 info_9 info_10 info_11  
132973    NaN    NaN    NaN     NaN     NaN  
132974    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

#### Observations:
- The `info` value is successfully divided and we can proceed through it column by column.
- We will check the set of values for the first two columns, for age.

### `info_0`

In [20]:
# Checking unique value counts
df["info_0"].value_counts()

                                                                       132946
Sir                                                                         6
92                                                                          2
Douglas Scott                                                               1
VC                                                                          1
Sir Woolwich West                                                           1
Sir Lord Justice of Appeal                                                  1
Dame MEP                                                                    1
83                                                                          1
Sir Governor-General                                                        1
Notable ice hockey players and coaches among the 44 killed in the :         1
Mike Alexander                                                              1
Colonel                                                         

<IPython.core.display.Javascript object>

#### Observations:
- The vast majority of rows have an empty string for this field.
- There is one row representing a group, rather than an individual, and we will drop it.
- We should verify the name and age information for the remainder of unique values in `info_0`.

#### Dropping Entry for Group

In [21]:
# Checking the entry representing a group
group_entry = df[
    df["info_0"]
    == "Notable ice hockey players and coaches among the 44 killed in the :"
]
group_entry

day                                  name  \
51130   7  2011 Lokomotiv Yaroslavl plane crash   

                                                                      info  \
51130  Notable ice hockey players and coaches among the 44 killed in the :   

                                                                     link  \
51130  https://en.wikipedia.org/wiki/2011_Lokomotiv_Yaroslavl_plane_crash   

      num_references  year      month  \
51130             95  2011  September   

                                                                    info_0  \
51130  Notable ice hockey players and coaches among the 44 killed in the :   

      info_1 info_2 info_3 info_4 info_5 info_6 info_7 info_8 info_9 info_10  \
51130    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   

      info_11  
51130     NaN

<IPython.core.display.Javascript object>

In [22]:
# Dropping group entry, resetting index, and checking new shape of df
df.drop(group_entry.index, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132974, 19)

<IPython.core.display.Javascript object>

#### Examining Rows with Atypical `info_0` Values

In [23]:
# Examining rows with atypical info_0 values
list_to_check = df["info_0"].value_counts().index.to_list()

verify_df = pd.DataFrame()
for item in list_to_check[1:]:
    verify_df = pd.concat([verify_df, df[df["info_0"] == item]])
verify_df

day                     name  \
30349   25           Julian Bullard   
31140    1          Kyffin Williams   
34207   15             Jeremy Moore   
43606   29         Derek Hodgkinson   
67212    7             Richard Best   
67217    7             Thomas Hinde   
33086   28           David Turnbull   
117387  11         Gotthilf Fischer   
67517   21             Colin Turner   
67162    5               Robin Dunn   
67152    4    Elaine Kellett-Bowman   
59616   24      Kristján Jóhannsson   
52580   25           Moses Pitakaka   
42140    5                    Evile   
40355   22          Dudley Thornton   
34171    9       Sir Tasker Watkins   
33751   22             John Burnett   
30468    8           Peter Smithers   
14035    5                slapstick   
30363   26       Mahmoud al-Majzoub   
30358   26  Horondino José da Silva   
30210    6        Flight Lieutenant   
30188    2         Sam Mokuahi, Jr.   
29619   10             James Yancey   
18908   11              were killed   
18312   20            Demetreus Nix   
15873   15                      NRG   
67661   30            David Gibbons   

                                                                                                                 info  \
30349                                                                                     Sir , 78, British diplomat.   
31140                                                               Sir , 88, Welsh artist, lung and prostate cancer.   
34207                                    Sir , 79, British soldier, commander of UK land forces in the Falklands War.   
43606                                                                            Sir , 92, British air chief marshal.   
67212                                                             Sir , 80, British diplomat, Ambassador to Iceland .   
67217                                                                                     Sir , 88, British novelist.   
33086                                                                             . 92, American materials scientist.   
117387                                                                                  92, German choral conductor .   
67517                                                            Sir Woolwich West, 92, British politician, MP for  .   
67162                                                              Sir Lord Justice of Appeal, 96, British jurist,  .   
67152                                              Dame MEP, 90, British politician and barrister, MP for Lancaster .   
59616                                                                                  83, Icelandic Olympic athlete.   
52580                                                       Sir Governor-General, 66, Solomon Islander politician,  .   
42140                                                       Mike Alexander, 32, British bassist , pulmonary embolism.   
40355                                                                             Colonel , 89, British army officer.   
34171   VC, 88, British jurist and businessman, Lord Justice of Appeal and former WRU President, after short illness.   
33751                                                 Sir Edinburgh University, 85, British academic, Principal of  .   
30468                 Sir Winchester, 92, British politician, MP for Secretary General of the Council of Europe and .   
14035                                Herbert Wiere, 91, Austrian-born American Wiere Brothers comedian, member of the   
30363                                  aka Abu Hamza, 41, Palestinian Islamic Jihad leader, assassination by bombing.   
30358                                     aka "Dino Sete Cordas", 88, Brazilian virtuouso of the seven-string guitar.   
30210                           Sarah-Jayne Mulvihill, 32, first British servicewoman to be killed in action in Iraq.   
30188     aka "Sammy Steamboat", 71, American Hawaiian professional wrestler, complications 

<IPython.core.display.Javascript object>

#### Observations:
- The majority of rows contain additional aliases or titles within `info_0`, that we don't need, but we can leave in place for now.  
- There are a few rows that will need to be treated individually to correct the name value, as follows:
    1. Entry is for Mike Alexander whose band was Evile.
    2. Entry is for Herbert Wiere who performed slapstick.
    3. Entry is for Sarah-Jayne Mulvihill who was a Flight Lieutenant.
    4. Entry is for Douglass Scott who was killed by Demetreus Nix.
    5. Entry is for Kim Hwan-Sung who was a member of the band NRG.
- We can replace the `name` value with the `info_0` value for these rows as well as proceed with hard-coding the correct values for info_2 and info_3 fields to match the Wikipedia pattern, but staying true to the information scraped.
- The row with "Nearly 3" value for `info_0` represents a group, rather than an individual, so will be dropped, after treating the above rows.
- We can proceed to extract age from `info_0` for the few rows that contain it here instead of in `info_1`.

#### Treating 5 rows with Name in `info_0`

In [24]:
values_lst = [
    "Mike Alexander",
    "Herbert Wiere",
    "Sarah-Jayne Mulvihill",
    "Douglas Scott",
    "Kim Hwan-Sung",
]

<IPython.core.display.Javascript object>

In [25]:
# For loop to copy name from info_0 to name
for i in df[df["info_0"].isin(values_lst)].index.to_list():
    df.loc[i, "name"] = df.loc[i, "info_0"]

# Hard-coding info_2 and info_3 values for Kim Hwan-Sung
index = df[
    df["link"] == "https://en.wikipedia.org/wiki/NRG_(South_Korean_band)"
].index.to_list()
df.loc[index, "info_2"] = "South Korean musician"

df.loc[index, "info_3"] = "respiratory illness"

# # Hard-coding info_2 and info_3 values for Douglass Scott
index = df[
    df["link"]
    == "https://en.wikipedia.org/w/index.php?title=Demetreus_Nix&action=edit&redlink=1"
].index.to_list()
df.loc[index, "info_2"] = "student"

df.loc[index, "info_3"] = "murdered"

# # Hard-coding info_2 and info_3 values for Sarah-Jayne Mulvihill
index = df[
    df["link"] == "https://en.wikipedia.org/wiki/Flight_Lieutenant"
].index.to_list()
df.loc[index, "info_2"] = "British servicewoman"

df.loc[index, "info_3"] = "killed in action"

<IPython.core.display.Javascript object>

In [26]:
df[df["info_0"].isin(values_lst)]

day                   name  \
14035   5          Herbert Wiere   
15873  15          Kim Hwan-Sung   
18312  20          Douglas Scott   
30210   6  Sarah-Jayne Mulvihill   
42140   5         Mike Alexander   

                                                                                        info  \
14035       Herbert Wiere, 91, Austrian-born American Wiere Brothers comedian, member of the   
15873                                                       Kim Hwan-Sung, 19, A Member of .   
18312                                   Douglas Scott, 20, High-school student murdered by .   
30210  Sarah-Jayne Mulvihill, 32, first British servicewoman to be killed in action in Iraq.   
42140                              Mike Alexander, 32, British bassist , pulmonary embolism.   

                                                                                 link  \
14035                                         https://en.wikipedia.org/wiki/Slapstick   
15873                           https://en.wikipedia.org/wiki/NRG_(South_Korean_band)   
18312  https://en.wikipedia.org/w/index.php?title=Demetreus_Nix&action=edit&redlink=1   
30210                                 https://en.wikipedia.org/wiki/Flight_Lieutenant   
42140                                             https://en.wikipedia.org/wiki/Evile   

      num_references  year    month                 info_0 info_1  \
14035             16  1999   August          Herbert Wiere     91   
15873             27  2000     June          Kim Hwan-Sung     19   
18312              0  2001     June          Douglas Scott     20   
30210             12  2006      May  Sarah-Jayne Mulvihill     32   
42140             82  2009  October         Mike Alexander     32   

                                               info_2               info_3  \
14035  Austrian-born American Wiere Brothers comedian        member of the   
15873                           South Korean musician  respiratory illness   
18312                                         student             murdered   
30210                            British servicewoman     killed in action   
42140                                 British bassist   pulmonary embolism   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  
14035    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
15873    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
18312    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
30210    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
42140    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

#### Dropping Entry for Group

In [27]:
# Checking the entry representing a group
group_entry = df[df["info_0"] == "Nearly 3"]
group_entry

day         name  \
18908  11  were killed   

                                                               info  \
18908  Nearly 3,000 people September 11 attacks in the , including:   

                                                                       link  \
18908  https://en.wikipedia.org/wiki/Casualties_of_the_September_11_attacks   

      num_references  year      month    info_0  \
18908            176  2001  September  Nearly 3   

                                       info_1      info_2 info_3 info_4  \
18908  000 people September 11 attacks in the  including:    NaN    NaN   

      info_5 info_6 info_7 info_8 info_9 info_10 info_11  
18908    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

In [28]:
# Dropping group entry, resetting index, and checking new shape of df
df.drop(group_entry.index, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132973, 19)

<IPython.core.display.Javascript object>

#### Extracting `age` from `info_0`

In [29]:
# For loop to extract age from info_0
for i, item in enumerate(df["info_0"]):
    pattern = r"\d"
    if re.search(pattern, item):
        df.loc[i, "age"] = int(item)

# Checking one of these rows
df[df["info_0"] == "92"]

day              name                                 info  \
33085   28    David Turnbull  . 92, American materials scientist.   
117386  11  Gotthilf Fischer        92, German choral conductor .   

                                                                      link  \
33085   https://en.wikipedia.org/wiki/David_Turnbull_(materials_scientist)   
117386                      https://en.wikipedia.org/wiki/Gotthilf_Fischer   

       num_references  year     month info_0                        info_1  \
33085               8  2007     April     92  American materials scientist   
117386              4  2020  December     92       German choral conductor   

       info_2 info_3 info_4 info_5 info_6 info_7 info_8 info_9 info_10  \
33085     NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   
117386    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   

       info_11   age  
33085      NaN  92.0  
117386     NaN  92.0

<IPython.core.display.Javascript object>

In [30]:
# Dropping info_0
df.drop("info_0", axis=1, inplace=True)

<IPython.core.display.Javascript object>

#### Observations:
- The new `age` column has been added successfully.
- We are finished processing `info_0`.

### `info_1`

#### Unique Values

In [31]:
# Checking unique values
df["info_1"].unique()

array(['86', '68', '87', '93', '79', '50', '88', '72', '81', '80', '90',
       '85', '92', '58', '54', '96', '49', '77', '76', '43', '35', '83',
       '31', '64', '57', '52', '84', '78', '70', '73', '67', '99', '33',
       '75', '66', '74', '62', '61', '82', '38', '47', '56', '91', '89',
       '94', '45', '65', '97', '63', '69', '37', '53', '46', '60', '26',
       '71', '25', '59', '39–40', '23', '95', '42', '32', '51', '41',
       '55', '44', '98', '39', '100', '27', '28', '40', '30', '48', '34',
       '29', '36', '111', '22', '104', '14', '21', '106', '105', '18',
       '101', '102', '20', '73-74', '84-85', '67-68', '85-86', '89-90',
       '103', '49-50', '87-88', '107', '48-49', '19', '71-72', '2',
       '82-83', '88-89', '11-12', '63-64', '64-65', '24', '17', '69-70',
       '52-53', '39-40', '80-81', '51-52', '70–71', '66–67', '59-60',
       '94-95', '15', '108', '76-77', '1995', '37-38',
       'American politician', '55/56', '87–88',
       '90 Swedish Olympic sprinte

<IPython.core.display.Javascript object>

#### Observations:
- There is a lot of variety in the format of the age data.
- Also, this field contains several values that we would expect in info_2 and beyond.
- Let us take the approach of extracting age values first, then examining rows with missing `age`.

#### Examining unique formats for age data

In [32]:
# Examining formats for age data
has_num = []
for i, item in enumerate(df["info_1"]):
    pattern = r"\d"
    if re.search(pattern, item) is not None:
        has_num.append(i)

df.loc[has_num, :]["info_1"].unique()

array(['86', '68', '87', '93', '79', '50', '88', '72', '81', '80', '90',
       '85', '92', '58', '54', '96', '49', '77', '76', '43', '35', '83',
       '31', '64', '57', '52', '84', '78', '70', '73', '67', '99', '33',
       '75', '66', '74', '62', '61', '82', '38', '47', '56', '91', '89',
       '94', '45', '65', '97', '63', '69', '37', '53', '46', '60', '26',
       '71', '25', '59', '39–40', '23', '95', '42', '32', '51', '41',
       '55', '44', '98', '39', '100', '27', '28', '40', '30', '48', '34',
       '29', '36', '111', '22', '104', '14', '21', '106', '105', '18',
       '101', '102', '20', '73-74', '84-85', '67-68', '85-86', '89-90',
       '103', '49-50', '87-88', '107', '48-49', '19', '71-72', '2',
       '82-83', '88-89', '11-12', '63-64', '64-65', '24', '17', '69-70',
       '52-53', '39-40', '80-81', '51-52', '70–71', '66–67', '59-60',
       '94-95', '15', '108', '76-77', '1995', '37-38', '55/56', '87–88',
       '90 Swedish Olympic sprinter', '6', '86-87', '62/63', '79

<IPython.core.display.Javascript object>

#### Observations:
- There are some specific rows that likely need to be dropped, with the following values for `info_1`:
    - 1995
    - 1997
    - German Olympic sailor [1]
    - Taiwanese failed assassin in the 3-19 shooting incident
    - 255
    - 176
    - the first wild bear in Germany in 170 years
    - c. 3500
    - common chimpanzee 55
    - Maltese 15
    - c.1000
    - Tree of the Year 150
- The dataset contains values for non-human species, so we will be on the lookout to drop those rows.
- The age formatting patterns include "age", "age1-age2" (short dash), "age1–age2" (long dash), 'age1/age2", "age ", 'age.', 'c. age', 'age/age 2nd digit', 'age or age', 'age?', 'early ages', 'age days', 'age-months', 'c.age', "age+", and "age months".
- We will need to address ages in days and months prior to those in years.

#### Examining Rows with Digits and Atypical Values for `info_2`

In [33]:
# List of atypical info_1 values for rows with digits
values_lst = [
    "1995",
    "1997",
    "German Olympic sailor [1]",
    "Taiwanese failed assassin in the 3-19 shooting incident",
    "255",
    "176",
    "the first wild bear in Germany in 170 years",
    "c. 3500",
    "common chimpanzee 55",
    "Maltese 15",
    "c.1000",
    "Tree of the Year 150",
]

df[df["info_1"].isin(values_lst)]

day                name  \
6046     9        Khaptad Baba   
6985     5  Wymberley D. Coerr   
10548   21       Pakoda Kadhar   
22722   17    Klaus Oldendorff   
25010   29      Chen Yi-hsiung   
29898   23             Adwaita   
30566   23             Harriet   
30591   26            Bear JJ1   
52996   16         The Senator   
55404    2              Oliver   
55455    5       Lucky Diamond   
86515    8  Pioneer Cabin Tree   
118228   5      Happy Man Tree   

                                                                                                                                       info  \
6046                                                                                                      , 1995, Nepalese spiritual saint.   
6985                                                                                              , 1995, American politician and diplomat.   
10548                                                                                                                 , 1997, Indian actor.   
22722                                                                                                           , German Olympic sailor [1]   
25010                                                                            , Taiwanese failed assassin in the 3-19 shooting incident.   
29898                          , 255 , tortoise claimant for world's oldest animal, reputedly a former pet of General Clive, liver failure.   
30566   , 176, Galápagos tortoise believed to be the third oldest animal in the world and allegedly owned by Charles Darwin, heart failure.   
30591                                                                         , the first wild bear in Germany in 170 years, shot to death.   
52996                                                                    , c. 3500, American pond cypress tree, largest in the world, fire.   
55404                                           , common chimpanzee 55, Congolese-born  noted for his upright stature and humanlike traits.   
55455                                                                   , Maltese 15, American Guinness World Record dog,  holder , cancer.   
86515                                                                                  , c.1000, American giant sequoia tree, storm damage.   
118228                                                                             , Tree of the Year 150, British plane tree,  , cut down.   

                                                     link num_references  \
6046           https://en.wikipedia.org/wiki/Khaptad_Baba              6   
6985     https://en.wikipedia.org/wiki/Wymberley_D._Coerr              3   
10548         https://en.wikipedia.org/wiki/Pakoda_Kadhar              1   
22722      https://en.wikipedia.org/wiki/Klaus_Oldendorff              1   
25010        https://en.wikipedia.org/wiki/Chen_Yi-hsiung              4   
29898               https://en.wikipedia.org/wiki/Adwaita              4   
30566    https://en.wikipedia.org/wiki/Harriet_(tortoise)             10   
30591              https://en.wikipedia.org/wiki/Bear_JJ1              1   
52996    https://en.wikipedia.org/wiki/The_Senator_(tree)             11   
55404   https://en.wikipedia.org/wiki/Oliver_(chimpanzee)             14   
55455   https://en.wikipedia.org/wiki/Lucky_Diamond_(dog)             13   
86515    https://en.wikipedia.org/wiki/Pioneer_Cabin_Tree             45   
118228       https://en.wikipedia.org/wiki/Happy_Man_Tree             23   

        year    month  \
6046    1996      May   
6985    1996  October   
10548   1998  January   
22722   2003    March   
25010   2004    March   
29898   2006    March   
30566   2006     June   
30591   2006     June   
52996   2012  January   
55404   2012     June   
55455   2012     June   
86515   2017  January   
118228  2021  January   

                                                         info_1  \
6046                                        

<IPython.core.display.Javascript object>

#### Observations:
- Age data is either missing or the entry is for a member of a non-human species.
- We will drop all of these rows.

#### Dropping Rows for Non-Human Entries or Entries Missing Age Data

In [34]:
# List of indexes to be dropped
drop_rows = df[df["info_1"].isin(values_lst)].index.to_list()

# Dropping rows, resetting index, and checking new shape of df
df.drop(drop_rows, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132960, 19)

<IPython.core.display.Javascript object>

### Extracting `age` from `info_1`

- Recall that the age formatting patterns include "age", "age1-age2" (short dash), "age1–age2" (long dash), 'age1/age2", "age ", 'age.', 'c. age', 'age/age 2nd digit', 'age or age', 'age?', 'early ages', 'age days', 'age-months', 'c.age', 'age+', and "age months".
- We should address age in days and months first.

#### Step 1: Age in Years and Months

In [35]:
# Dictionary of patterns for days and months formats as keys and factor to convert to years
patterns = {
    r"(\d{1,3})( days)": 365,
    r"(\d{1,3})(-months)": 12,
    r"(\d{1,3})( months)": 12,
}

# For loop to check for age in days and months and convert to years and save in age.
# Also removes age in days/months from info_0.
for key, value in patterns.items():
    for i, item in enumerate(df["info_1"]):
        match = re.search(key, item)
        if match:
            age = int(match.group(1)) / value
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(key, "", df.loc[i, "info_1"])

# Checking updated rows
df[df["age"].notna()]

day                         name  \
28492   12  Susan Anne Catherine Torres   
30524   18          Chris and Cru Kahui   
33077   28               David Turnbull   
59604   24          Kristján Jóhannsson   
90456   28                 Charlie Gard   
117374  11             Gotthilf Fischer   

                                                                                                            info  \
28492   , 40 days, American baby born to Susan Torres, brain-dead woman, heart failure after intestinal surgery.   
30524                                                            , 3-months, New Zealand child homicide victims.   
33077                                                                        . 92, American materials scientist.   
59604                                                                             83, Icelandic Olympic athlete.   
90456                       , 11 months, British infant, subject of life support and parental rights case, MDDS.   
117374                                                                             92, German choral conductor .   

                                                                         link  \
28492                              https://en.wikipedia.org/wiki/Susan_Torres   
30524                       https://en.wikipedia.org/wiki/Chris_and_Cru_Kahui   
33077      https://en.wikipedia.org/wiki/David_Turnbull_(materials_scientist)   
59604   https://en.wikipedia.org/wiki/Kristj%C3%A1n_J%C3%B3hannsson_(athlete)   
90456                         https://en.wikipedia.org/wiki/Charlie_Gard_case   
117374                         https://en.wikipedia.org/wiki/Gotthilf_Fischer   

       num_references  year      month                        info_1  \
28492               3  2005  September                                 
30524              48  2006       June                                 
33077               8  2007      April  American materials scientist   
59604               2  2013    January     Icelandic Olympic athlete   
90456             125  2017       July                                 
117374              4  2020   December       German choral conductor   

                                    info_2  \
28492   American baby born to Susan Torres   
30524   New Zealand child homicide victims   
33077                                  NaN   
59604                                  NaN   
90456                       British infant   
117374                                 NaN   

                                                  info_3  \
28492                                   brain-dead woman   
30524                                                NaN   
33077                                                NaN   
59604                                                NaN   
90456   subject of life support and parental rights case   
117374                                               NaN   

                                        info_4 info_5 info_6 info_7 info_8  \
28492   heart failure after intestinal surgery    NaN    NaN    NaN    NaN   
30524                                      NaN    NaN    NaN    NaN    NaN   
33077                                      NaN    NaN    NaN    NaN    NaN   
59604                                      NaN    NaN    NaN    NaN    NaN   
90456                                     MDDS    NaN    NaN    NaN    NaN   
117374                                     NaN    NaN    NaN    NaN    NaN   

       info_9 info_10 info_11        age  
28492     NaN     NaN     NaN   0.109589  
30524     NaN     NaN     NaN   0.250000  
33077     NaN     NaN     NaN  92.000000  
59604     NaN     NaN     NaN  83.000000  
90456     NaN     NaN     NaN   0.916667  
117374    NaN     NaN     NaN  92.000000

<IPython.core.display.Javascript object>

#### Observations:
- We have successfully captured the age in days and months values and converted them to years.
- The other rows are in place already from our treatment of `info_0`.
- Next, we will extract the entries with straightforward single integer age values, including the formats: "age", "age ", 'age.'.
- Apparent estimates are excluded here to allow closer examination, as they are more likely to by atypical entries.

#### Step 2: Age as Single Integer (Excluding Estimates)

In [36]:
# List of patterns for age formats with single integer for age
patterns = [r"^(\d{1,3})$", r"^(\d{1,3})\s", r"^(\d{1,3})\.\s"]

# For loop to check age pattern in info_0, save age to age column, and remove from age from info_0
for i, item in enumerate(df["info_1"]):
    for pattern in patterns:
        match = re.search(pattern, item)
        if match:
            age = int(match.group(1))
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])

# Checking first 2 rows
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_1           info_2                        info_3  \
0  1994  January          British dancer  ballet designer and director   
1  1994  January         Irish economist                        writer   

         info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  
0           NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  86.0  
1  and academic    NaN    NaN    NaN    NaN    NaN     NaN     NaN  68.0

<IPython.core.display.Javascript object>

In [37]:
# Checking last 2 rows
df.tail(2)

day           name  \
132958   9  Oleg Moliboga   
132959   9       Zou Jing   

                                                                         info  \
132958          , 69, Russian volleyball player, Olympic champion  and coach.   
132959  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132958        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132959  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_1                     info_2  \
132958  2022  June         Russian volleyball player   
132959  2022  June                  Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132958                   Olympic champion  and coach    NaN    NaN    NaN   
132959  member of the Chinese Academy of Engineering    NaN    NaN    NaN   

       info_7 info_8 info_9 info_10 info_11   age  
132958    NaN    NaN    NaN     NaN     NaN  69.0  
132959    NaN    NaN    NaN     NaN     NaN  86.0

<IPython.core.display.Javascript object>

In [38]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

1014

<IPython.core.display.Javascript object>

#### Observations:
- The rows with single integer age data have been addressed.
- There are 1014 remaining missing values for `age`.
- Let us check the rows containing 'c.' in the age information.

#### Entries with Age Data Containing 'c.', '+', or '?' for Estimate

In [39]:
values = ["c.", "+", "?"]

# List of rows to check
rows_to_check = []

# For loop to add index of list containing values to list
for index in df[df["age"].isna()].index:
    if any(value in df.loc[index, "info_1"] for value in values):
        rows_to_check.append(index)

# Inspecting rows containing values
df.loc[rows_to_check, :]

day                                  name  \
12922   22                         Graciela Quan   
25366    4                   Tesfaye Gebre Kidan   
25446   18                          Paul Johnson   
25447   18                          Nek Mohammed   
26727   13                          Earl Cameron   
29744   28                            Peter Snow   
29840   15                              Humphrey   
30259   14                            Paul Marco   
31174    6          Mohammed Taha Mohammed Ahmed   
31456   11                       Benito Martínez   
32768   21                       Mohd. Ayub Khan   
33117    5     Prince Abdul-Majid bin Abdul-Aziz   
33990   21                             Čabulītis   
34572   30                                Washoe   
34626    6                  Sayed Mustafa Kazemi   
38030    1                        Arlene Sherman   
39780    2                         Halit Balamir   
40778    3                           Moloko Temo   
41420    4                                Benson   
41439    5  Sheikha Hessa bint Salman Al Khalifa   
41440    5                      Baitullah Mehsud   
41896   14                Saleh Ali Saleh Nabhan   
43570   28              Mohammad-Reza Ali-Zamani   
43588   28                     Arash Rahmanipour   
44022   27                     Black Bear Island   
44099    4                             Big Truck   
44114    5                        Aminah Assilmi   
45203   24                       Abdolhamid Rigi   
47376   15                        Nimr al-Khatib   
48567   18                             Bob Tanna   
49231   12                         Buster Martin   
49491    1                            Anny Rüegg   
49984    8               Fazul Abdullah Mohammed   
53783   28                Abukar Hassan Mohamoud   
54952    5                                  Meow   
55771   24                       Lonesome George   
56360   31           Abdi Jeylani Malaq Marshale   
58606    9                              Ataa Oko   
61245   17                             Bi Kidude   
61387   23                         Mohammed Omar   
62113    4                           Luo Meizhen   
62547   27                      Muhammad Emin Er   
64751    3                                Reshma   
66410   30                               Greater   
67512   22                    Hamza Abu al-Haija   
69858   22                          Glenn Jowitt   
70433   21                          Raed al Atar   
71116   26                                 Zelda   
73701    4      Sajida Mubarak Atrous al-Rishawi   
74507   19             Leandra Becerra Lumbreras   
74711   28                         Tuti Yusupova   
75567   13                           Bob Randall   
75797   27        Hassan Abdullah Hersi al-Turki   
78509    2                        Omar El-Hariri   
82378   21                        Akhtar Mansour   
92359   14                           Little Mama   
95096   21                            Kakon Bibi   
100116  17                        Zura Karuhimbi   
105359  19                        Carlos Porrata   
109858  13                        Nasser Shabani   
128545  20                     Mamadú Iaia Djaló   

                                                                                                                           info  \
12922                                                                   , c. 79, Guatemalan lawyer and women's rights activist.   
25366                                                   , c. 69, Ethiopian general, defense minister and President of Ethiopia.   
25446                                                                       , c. 49, American hostage, decapitated by al-Qaeda.   
25447                 , c. 27, Pakistani tribal leader in Waziristan and key Taliban ally, killed by Pakistani military forces.   
26727                                                                                 , 89?, Canadian broadcaster and  anchor .

<IPython.core.display.Javascript object>

#### Observations:
- Most of the entries are for people, but there are also one or more entries for the following:
    - carp
    - racehorse
    - chimpanzee
    - flamingo
    - cat
    - turkey
- We will proceed to check `info_2` for these values and drop these and other rows representing members of these other species.

#### Checking for Cat, Racehorse, Chimpanzee, Carp, and Flamingo in `info_2`

In [40]:
# Defining pattern for re
species_pattern = r"\b(cat|racehorse|chimpanzee|carp|flamingo|turkey)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Checking number of rows containing pattern
print(f"There are {len(rows_to_check)} rows containing these values.")

There are 468 rows containing these values.


<IPython.core.display.Javascript object>

#### Observations:
- There are sufficient rows to warrant checking species by species.

#### Cat Entries per `info_2`

In [41]:
# Defining pattern for re
species_pattern = r"\b(cat)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :]

day                  name  \
18611    4         Lorenzo Music   
31856   29  Dewey Readmore Books   
38128   11              Scarlett   
39071    4                 India   
39689   20                 Socks   
41344   27                 Sybil   
47437   21          Prince Chunk   
54952    5                  Meow   
59825    4                Stewie   
63272    7     Buurtpoes Bledder   
66390   29          Colonel Meow   
70838   11       Hamish McHamish   
72424    4       Frank and Louie   
76284   22                  Tama   
90322   21                Stubbs   
92238    7               Paddles   
103399  14            Grumpy Cat   
107383   1               Lil Bub   
109614   3                    CC   

                                                                                                                              info  \
18611   , 64, American voice actor known for the voice of the cartoon cat Garfield, complications related to lung and bone cancer.   
31856                                                                                               , 19,  Library cat, euthanized   
38128                           , 13, American stray cat, name source of the Scarlett Award for Animal Heroism, animal euthanasia.   
39071                                                                                    , 18, American pet cat of George W. Bush.   
39689                                                           , 19, American Presidential cat of the Clinton family, euthanized.   
41344                                   , 3, British Downing Street cat, Chief Mouser to the Cabinet Office , after short illness.   
47437                                                                                     , 10, American obese cat, heart disease.   
54952                                                       , c. 2, American cat, heaviest cat at his time of death, lung failure.   
59825                                                                                 , 7–8, world's longest domestic cat, cancer.   
63272                                                                          , 1–2, domestic cat in Netherlands, motor accident.   
66390                                           , 2, American Himalayan-Persian cat, Guinness World Record holder for longest fur.   
70838                                                                               , 15, Scottish celebrity cat, chest infection.   
72424                                                  , 15, American-born Ragdoll cat, world's oldest diprosopus cat, euthanized.   
76284                                                    , 16, Japanese calico cat, stationmaster at Kishi Station, heart failure.   
90322                                                                    , 20, American cat, honorary mayor of Talkeetna, Alaska .   
92238                               , 1, New Zealand polydactyl cat, co-owned by Prime Minister Jacinda Ardern, traffic collision.   
103399                                           , 7, American internet celebrity cat, complications from urinary tract infection.   
107383                                                                                               , 8, American celebrity cat .   
109614                                                                 , 18, American tabby cat, first cloned pet, kidney failure.   

                                                      link num_references  \
18611          https://en.wikipedia.org/wiki/Lorenzo_Music              6   
31856   https://en.wikipedia.org/wiki/Dewey_Readmore_Books             26   
38128         https://en.wikipedia.org/wiki/Scarlett_(cat)              6   
39071            https://en.wikipedia.org/wiki/India_(cat)              7   
39689            https://en.wikipedia.org/wiki/Socks_(cat)             21   
41344            https://en.wikipedia.org/wiki/Sybil_(cat)              5   
47437           https://en.wikipedia.org/wiki/Prince_Chunk        

<IPython.core.display.Javascript object>

#### Observations:
- There is one person represented.  
- We can proceed to drop the others, using "cat", "cat of", and "cat in" patterns.

#### Dropping Entries for Cats per `info_2`

In [42]:
# List of re patterns to find
patterns = [r"\bcat$", r"\b(cat of|cat in)\b"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
print(f"{len(rows_to_drop)} rows will be dropped.")

18 rows will be dropped.


<IPython.core.display.Javascript object>

In [43]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132942, 19)

<IPython.core.display.Javascript object>

####  Racehorse Entries per `info_2`

In [44]:
# Defining pattern for re
species_pattern = r"\bracehorse\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :].sample(10)

day              name  \
12071    2    Tommy J. Smith   
88651   21        Cape Cross   
121975  24  John T. Ward Jr.   
104973  30       Deep Impact   
56568   13     Typhoon Tracy   
71899    7           Solwhit   
114845  29           Subzero   
75243   24   Success Express   
37304   22      Ballindaggin   
87897   13   Danehill Dancer   

                                                                                 info  \
12071                                             , 81, Australian racehorse trainer.   
88651                                , 23, Irish-bred racehorse and sire, euthanized.   
121975                                              , 75, American racehorse trainer.   
104973                       , 17, Japanese champion racehorse  and sire, euthanised.   
56568        , 6, Australian Thoroughbred racehorse, winner of the Coolmore Classic .   
71899    , 10, French-bred Irish-trained Thoroughbred racehorse, fall while training.   
114845                 , 31, Australian racehorse, Melbourne Cup winner , euthanized.   
75243                                          , 30, American Thoroughbred racehorse.   
37304                              , 23, American Thoroughbred racehorse, euthanized.   
87897   , 24, Irish-bred British-trained thoroughbred racehorse and sire, euthanized.   

                                                      link num_references  \
12071         https://en.wikipedia.org/wiki/Tommy_J._Smith             14   
88651     https://en.wikipedia.org/wiki/Cape_Cross_(horse)              3   
121975      https://en.wikipedia.org/wiki/John_T._Ward_Jr.              1   
104973   https://en.wikipedia.org/wiki/Deep_Impact_(horse)              3   
56568          https://en.wikipedia.org/wiki/Typhoon_Tracy              5   
71899                https://en.wikipedia.org/wiki/Solwhit              8   
114845       https://en.wikipedia.org/wiki/Subzero_(horse)              8   
75243        https://en.wikipedia.org/wiki/Success_Express              1   
37304   https://en.wikipedia.org/wiki/Ballindaggin_(horse)              9   
87897        https://en.wikipedia.org/wiki/Danehill_Dancer             34   

        year      month info_1  \
12071   1998  September          
88651   2017      April          
121975  2021      April          
104973  2019       July          
56568   2012     August          
71899   2014   November          
114845  2020     August          
75243   2015      April          
37304   2008       July          
87897   2017      March          

                                                            info_2  \
12071                                 Australian racehorse trainer   
88651                                Irish-bred racehorse and sire   
121975                                  American racehorse trainer   
104973                       Japanese champion racehorse  and sire   
56568                            Australian Thoroughbred racehorse   
71899             French-bred Irish-trained Thoroughbred racehorse   
114845                                        Australian racehorse   
75243                              American Thoroughbred racehorse   
37304                              American Thoroughbred racehorse   
87897   Irish-bred British-trained thoroughbred racehorse and sire   

                                info_3      info_4 info_5 info_6 info_7  \
12071                              NaN         NaN    NaN    NaN    NaN   
88651                       euthanized         NaN    NaN    NaN    NaN   
121975                             NaN         NaN    NaN    NaN    NaN   
104973                      euthanised         NaN    NaN    NaN    NaN   
56568   winner of the Coolmore Classic         NaN    NaN    NaN    NaN   
71899              fall while training         NaN    NaN    NaN    NaN   
114845            Melbourne Cup winner  euthanized    NaN    NaN    NaN   
75243                              NaN         NaN    NaN    NaN    NaN   
37304

<IPython.core.display.Javascript object>

#### Observations:
- There are several entries for people involved in the racehorse business.
- Values that end in 'racehorse' and 'racehorse and sire' can be removed.

#### Dropping Entries for Racehorses per `info_2`

In [45]:
# List of re patterns to find
patterns = [r"\bracehorse$", r"\b(racehorse and sire)$"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
len(rows_to_drop)

349

<IPython.core.display.Javascript object>

In [46]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132593, 19)

<IPython.core.display.Javascript object>

####  Chimpanzee, Flamingo,  Carp, and Turkey Entries per `info_2`

In [47]:
# Defining pattern for re
species_pattern = r"\b(chimpanzee|flamingo|carp|turkey)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :]

day                 name  \
34557  30               Washoe   
39599  16               Travis   
41365   4               Benson   
45231   1  Heather the Leather   
66269  30              Greater   
70953  26                Zelda   
76076  22     Don Featherstone   
92091  14          Little Mama   

                                                                                                                        info  \
34557  , c.42, American-trained African-born chimpanzee believed to be first non-human to acquire human language, influenza.   
39599                                                    , 13, American-born chimpanzee, television commercial animal, shot.   
41365                                                      , c. 25, British common carp, voted as Britain's Favourite Carp .   
45231                                                                                 , 50, British scaleless carp, old age.   
66269                                             , c. 83, Australian greater flamingo, world's oldest flamingo, euthanized.   
70953                              , 11+, American wild turkey, resident of New York City's Battery Park, traffic collision.   
76076                                   , 79, American artist and inventor of the plastic pink flamingo, Lewy body dementia.   
92091                                                    , c.79, African-born chimpanzee , oldest on record, kidney failure.   

                                                          link num_references  \
34557        https://en.wikipedia.org/wiki/Washoe_(chimpanzee)             42   
39599        https://en.wikipedia.org/wiki/Travis_(chimpanzee)             69   
41365              https://en.wikipedia.org/wiki/Benson_(fish)              9   
45231        https://en.wikipedia.org/wiki/Heather_the_Leather              5   
66269         https://en.wikipedia.org/wiki/Greater_(flamingo)             12   
70953             https://en.wikipedia.org/wiki/Zelda_(turkey)             13   
76076  https://en.wikipedia.org/wiki/Don_Featherstone_(artist)             10   
92091                https://en.wikipedia.org/wiki/Little_Mama              3   

       year      month info_1  \
34557  2007    October   c.42   
39599  2009   February          
41365  2009     August  c. 25   
45231  2010       June          
66269  2014    January  c. 83   
70953  2014  September    11+   
76076  2015       June          
92091  2017   November   c.79   

                                                                                                  info_2  \
34557  American-trained African-born chimpanzee believed to be first non-human to acquire human language   
39599                                                                           American-born chimpanzee   
41365                                                                                British common carp   
45231                                                                             British scaleless carp   
66269                                                                        Australian greater flamingo   
70953                                                                               American wild turkey   
76076                                          American artist and inventor of the plastic pink flamingo   
92091                                                                            African-born chimpanzee   

                                         info_3             info_4 info_5  \
34557                                 influenza                NaN    NaN   
39599              television commercial animal               shot    NaN   
41365         voted as Britain's Favourite Carp                NaN    NaN   
45231                                   old age                NaN    NaN   
66269                   world's oldest flamingo         euthanized    NaN   
70953  resident of New York City's Battery Park  traffic collision    NaN   
76076 

<IPython.core.display.Javascript object>

#### Observations:
- All of these rows are for animals, so we will remove them.

#### Dropping Entries for Chimpanzees, Flamingos,  Carps, and Turkeys per `info_2`

In [48]:
# List of re patterns to find
patterns = [r"\b(chimpanzee|flamingo|carp|turkey)\b"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
len(rows_to_drop)

8

<IPython.core.display.Javascript object>

In [49]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132585, 19)

<IPython.core.display.Javascript object>

#### Observations:
- With those non-human entries addressed, we can return to processing `info_`.
- Let us address the remaining entries with '?' and 'c.', accepting the estimated age as `age`.
- The sole entry containing '+' in the age estimate has been removed.

#### Extracting `age` from `info_1` for Entries with Age Estimate Containing '?' or 'c.'

In [50]:
# List to identify rows
values = ["c.", "?"]

# Pattern for re
pattern = r"\b(\d{1,3})\b"

# For loop to find rows with values and pattern and extract age to age column and remove age from info_1
for i in df[df["age"].isna()].index:
    item = df.loc[i, "info_1"]

    if any(value in item for value in values):
        match = re.search(pattern, item)

        if match:
            age = match.group(1)
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])

        for value in values:
            df.loc[i, "info_1"] = df.loc[i, "info_1"].replace(value, "")

# Checking example rows
pd.concat([df[df["name"] == "Tuti Yusupova"], df[df["name"] == "Raed al Atar"]])

day           name  \
74517  28  Tuti Yusupova   
70268  21   Raed al Atar   

                                                                            info  \
74517  , 134?, Uzbekistani longevity claimant, unverified world's oldest person.   
70268                            , c. 40, Palestinian militant , missile strike.   

                                              link num_references  year  \
74517  https://en.wikipedia.org/wiki/Tuti_Yusupova              5  2015   
70268   https://en.wikipedia.org/wiki/Raed_al_Atar             19  2014   

        month info_1                          info_2  \
74517   March         Uzbekistani longevity claimant   
70268  August                   Palestinian militant   

                                 info_3 info_4 info_5 info_6 info_7 info_8  \
74517  unverified world's oldest person    NaN    NaN    NaN    NaN    NaN   
70268                    missile strike    NaN    NaN    NaN    NaN    NaN   

      info_9 info_10 info_11  age  
74517    NaN     NaN     NaN  134  
70268    NaN     NaN     NaN   40

<IPython.core.display.Javascript object>

#### Observations:


In [51]:
df[df["age"].isna()]["info_1"].unique()

array(['39–40', '73-74', '84-85', '67-68', '85-86', '89-90', '49-50',
       '87-88', '48-49', '71-72', '82-83', '88-89', '11-12', '63-64',
       '64-65', '69-70', '52-53', '39-40', '80-81', '51-52', '70–71',
       '66–67', '59-60', '94-95', '76-77', '37-38', 'American politician',
       '55/56', '87–88', '86-87', '62/63', '79-80', '1996', '100/101',
       '97/98', '91/92', '66-67', '66/67', '38-39', '82–83', '88–89', '',
       '92–93', '78/79', 'Prime Minister of Zaire', '85/86', '54/55',
       '50/51', '73/74', '67/68', '45–46', '34/35', '69–70', '75/76',
       'King of Nepal', '89/90', '72/3',
       'member of the then National Assembly of Pakistan and Union Minister of Labor',
       '87/88', '28/29', '90/91', '52/53', '69/70', '73/4', '72-73',
       '52–53', '85–86', '82/83', 'Jules Engel', '94/95', '75-76', '76/7',
       '38-40', '60–61', '43/4', 'Vanuatuan president', '53/54', '55–56',
       '81-82', '54/5', 'early 80s', '93/4', '86/7', 'early 40s', '46/47',
       '8

<IPython.core.display.Javascript object>

In [ ]:
digits = [str(digit) for digit in np.arange(0, 10)]
rows_to_check = []
for index in df[df["age"].isna()].index:
    if any(digit in df.loc[index, "info_1"] for digit in digits):
        rows_to_check.append(index)
df.loc[rows_to_check, :].sample(10)

In [ ]:
df[df["info_1"] == "80. New Zealand Maori leader"]

In [ ]:
df[df["info_1"] == "11+"]